tests: benchmarks: timing_info: Add support for ARC, XTENSA, NIOS2, RISCV

AdithyaBaglody opened the pull request:
2 Parts to this PR. 

Part 1:
Added new KPIs for the userspace benchmarks.
-  Drop to user mode.
-  User thread Creation.
-  Syscall overhead.
-  Validation overhead k object init.
-  Validation overhead k object permission

Part 2: 
- Expand all possible KPIs to ARC, XTENSA, NIOS2 and RISCV32
- Provide userspace KPIs to only x86, ARM and ARC (the 3 architectures in which userspace is enabled)

AdithyaBaglody committed:
benchmarks: timing_info: Add hooks in the kernel for userspace.

Added sampling hooks in the kernel needed for userspace benchmarks.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

andrewboie commented on arch/x86/core/userspace.S:
```diff
@@ -169,6 +169,19 @@ SECTION_FUNC(TEXT, _x86_userspace_enter)
 	 */
 	mov	%esp, %edi
 
+#ifdef CONFIG_EXECUTION_BENCHMARKING
```
You might be able to get away with moving this right before the "iret" since the code before it just formats the stack.

AdithyaBaglody replied:
Sure I can move it.

andrewboie commented on include/misc/timing_info.h:
```diff
@@ -0,0 +1,24 @@
+__syscall u32_t userspace_read_timer_value(void);
+__syscall int validation_overhead_syscall(void);
+#endif  /* CONFIG_EXECUTION_BENCHMARKING */
+
```
You need an #include <syscalls/timing_info.h> at the bottom of this header.

andrewboie commented on include/misc/timing_info.h:
```diff
@@ -0,0 +1,24 @@
+/*
```
The syscalls defined here are specific to a single test case.
This should *not* go in include/misc/, this needs to be in the same codebase as the test case.
We might need to extend the syscall auto-magic to have additional search paths,

andrewboie commented on tests/benchmarks/timing_info/src/timing_info.h:
```diff
@@ -16,23 +16,58 @@
-#define TIMING_INFO_PRE_READ() (NRF_TIMER2->TASKS_CAPTURE[0] = 1)
-#define TIMING_INFO_OS_GET_TIME()  (NRF_TIMER2->CC[0])
-#define TIMING_INFO_GET_TIMER_VALUE()	TIMING_INFO_GET_CURRENT_TIME()
+#define TIMING_INFO_PRE_READ()        (NRF_TIMER2->TASKS_CAPTURE[0] = 1)
```
We really need some arch-independent APIs for getting highest-resolution timing information instead of ifdefing our test cases like this

andrewboie commented on tests/benchmarks/timing_info/src/timing_info.h:
```diff
@@ -16,23 +16,58 @@
-/* NRF RTC TIMER runs ar very slow rate (32KHz), So in order to measure
- * Kernel starts a dedicated timer to measure kernel stats.
+/* Stack size for all the threads created in this benchmark */
+#define STACK_SIZE 1024
```
`+ CONFIG_TEST_EXTRA_STACKSIZE` (or whatever it is)

andrewboie commented on tests/benchmarks/timing_info/prj_userspace.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_EXECUTION_BENCHMARKING=y
+CONFIG_TEST_RANDOM_GENERATOR=y
+CONFIG_HEAP_MEM_POOL_SIZE=256
+CONFIG_MAIN_STACK_SIZE=2048
```
why is main stacksize increased?

AdithyaBaglody replied:
actaully no specific reason. These configs are taken from the old prj.conf file.

andrewboie commented on tests/benchmarks/timing_info/prj_userspace.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_FORCE_NO_ASSERT=y
+
+#Disable Userspace
```
comment is opposite of what you are doing here. this is redundant anyway since CONFIG_TEST selects CONFIG_TEST_USERSPACE and CONFIG_TEST_HW_STACK_PROTECTION

andrewboie commented on tests/benchmarks/timing_info/src/userspace_bench.c:
```diff
@@ -0,0 +1,267 @@
+			-1 /*priority*/, K_INHERIT_PERMS | K_USER, 0);
+
+
+	u32_t total_cycles_obj_init = (u32_t)
```
I keep seeing lines like this, any way to macro it to make simpler?

andrewboie commented on tests/benchmarks/timing_info/testcase.yaml:
```diff
@@ -1,4 +1,9 @@
+  benchmark.timing.userspace:
+    filter: CONFIG_ARCH_HAS_USERSPACE
+    extra_args: CONF_FILE=prj_userspace.conf
+    arch_whitelist: x86 arm
```
how hard would it be to add ARC support?
just need to add another ifdef block for the timer right?

AdithyaBaglody replied:
currently adding the support. i will be creating another PR once that is done.

AdithyaBaglody committed:
arch: arc: Added benchmark related hooks.

The benchmark application timing_info needs certain hooks to be
present in the kernel to get the accurate measurements. This
patch adds these hook at all the required locations.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: benchmarks: timing_info: Enable benchmarks for ARC.

This patch provides support needed to get timing related
information from ARC based SOC.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: benchmarks: timing_info: Discard selected measurements.

Discard selected measurements that occur immediately before and
after the timer interrupt occurs. This causes fluctuations in the
time measured.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

andrewboie commented on tests/CMakeLists.txt:
```diff
@@ -3,3 +3,6 @@ add_subdirectory_if_kconfig(ztest)
   )
+
+zephyr_syscall_include_ifdef(CONFIG_EXECUTION_BENCHMARKING
+  "${CMAKE_CURRENT_SOURCE_DIR}/benchmarks/timing_info/src")
```
why is this here?
this should go in the CMakeLists.txt for the specific timing_info test!

AdithyaBaglody committed:
tests: benchmarks: timing_info: Enable benchmarks for xtensa.

This patch provides support needed to get timing related
information from xtensa based SOC.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: benchmarks: timing_info: Enable benchmarks for nios2.

This patch provides support needed to get timing related
information from nios2 based SOC.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: include: Add implementation of timestamp_serialize()

Add implementation of timestamp_serialize() for xtensa, nios2
and riscv. The implementation for the above architectures is an
empty function.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: benchmarks: timing_info: Add userspace related KPIs.

Added new KPIs for the userspace benchmarks.
1. Drop to user mode.
2. User thread Creation.
3. Syscall overhead.
4. Validation overhead k object init.
5. Validation overhead k object permission.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
tests: benchmarks: timing_info: Enable benchmarks for riscv32.

This patch provides support needed to get timing related
information from riscv32 based SOC.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

nashif commented on tests/benchmarks/timing_info/testcase.yaml:
```diff
@@ -1,4 +1,9 @@
-  benchmark.timing:
-    arch_whitelist: x86 arm
+  benchmark.timing.default_kernel:
+    arch_whitelist: x86 arm arc
```
shouldn't this whitelist more architectures, actually should run all?

AdithyaBaglody replied:
yeah the remaining architectures should come here. I missed it

AdithyaBaglody committed:
tests: benchmarks: timing_info: Cleanup testcase.yaml

The only architecture not supported is native posix. Added it to
arch_exclude.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody commented on tests/benchmarks/timing_info/testcase.yaml:
```diff
@@ -1,4 +1,9 @@
-  benchmark.timing:
-    arch_whitelist: x86 arm
+  benchmark.timing.default_kernel:
+    arch_whitelist: x86 arm arc xtensa nios2 riscv32
```
yep makes sense. Done.

nashif commented:
@AdithyaBaglody  please rebase. Did you test this on ARC? it fails for me using the nsim simulator:

```
***** Booting Zephyr OS v1.12.0-937-g9189109847 *****
starting test - Time Measurement
Timing Results: Clock Frequency: 10 MHz
Exception vector: 0x6, cause code: 0x2, parameter 0x4
Address 0xfffffffc
Current thread ID = 0x8000001c
Faulting instruction address = 0x29a4
Fatal fault in essential thread! Spinning...
```

nashif commented:
@AdithyaBaglody great, thanks, I just verified it on nsim_sem and emsk
